File: Task#2/C++/Miller.cpp
#include <iostream>
#include <vector>
#include <cstdlib>
#include <ctime>
#include <algorithm>
#include <string>
#include <set>
#include <cmath>
#include <map>
#include <iomanip>

using namespace std;

int powmod(int a, int x, int mod){
    int res = 1;
    for (int i = 1; i <= x; i++){
        res = res * a;
        res = res % mod;
    }
    return res;
}

vector<int> sieveEratosthenes(int n) {
    vector<bool> is_prime(n + 1, true);
    vector<int> primes;
    
    is_prime[0] = is_prime[1] = false; // 0 и 1 не являются простыми
    
    for (int i = 2; i * i <= n; i++) {
        if (is_prime[i]) {
            for (int j = i * i; j <= n; j += i) {
                is_prime[j] = false; // отмечаем все кратные i как составные
            }
        }
    }
    
    for (int i = 2; i <= n; i++) {
        if (is_prime[i]) {
            primes.push_back(i);
        }
    }
    
    return primes;
}

bool probability_Test(int num, int k){ // Миллер-Рабин
    if (num == 2 || num == 3){
        return true;
    }
    if (num < 2 || num % 2 == 0){
        return false;
    }
    int s = 0;
    int d = num-1;
    while (d % 2 == 0){
        d /= 2;
        s++;
    }
    for (int i = 0; i < k; i++){
        int a = rand() % (num - 4) + 2; // от 2 до num - 2
        int x = powmod(a, d, num);
        int y;
        for (int j = 0; j < s; j++){
            y = powmod(x, 2, num);
            if (y == 1 && x != 1 && x != (num - 1)) {
                return false;
            }
            x = y;
        }
        if (y != 1) {
            return false;
        }
    }
    return true;
}

int canon(int n, vector<int> A){
    vector<int> prime = sieveEratosthenes(500);
    vector<int> check;
    map<int, int> q;
    int N = n - 1;

    for (int i = 0; i < prime.size(); i++){
        int R = N;
        int k = 0;
        if (R % prime[i] == 0){
            while(R % prime[i] == 0){
                R /= prime[i];
                k += 1;
            }
            q.insert({prime[i], k});
        }
        else {
            continue;
        }
    }

    for (auto i : q) {
        for (int j = 0; j < A.size(); j++){
            int res = (powmod(j, ((n - 1)/i.first), n));
            check.push_back(res);
            if (res != 1){
                break;
            }
        }
    }

    for (int i = 0; i < check.size(); i++){
        if (check[i] != 1){
            return 1;
        }
    }

    return 0;
}

int test_Miller(int n, int t){

    vector<int> A;

    for (int i = 0; i < t; i++){
        A.push_back(rand() % (n - 2) + 2);
    }

    int k = 0;
    for (int i = 0; i < A.size(); i++){
        int res = powmod(A[i], n - 1, n);
        k += res;
    }

    if (A.size() != k){
        return false;
    }

    int res = canon(n, A);

    if (res == 1) {
        return true;
    }
    else {
        return false;
    }
}

void printTable(const vector<int>& primeNums, const vector<bool>& primeTestAgree, const vector<int>& primeLostCount) {
    int n = primeNums.size();
    int colWidth = 12;

    auto printLine = [&](int count) {
        cout << "+";
        for (int i = 0; i <= count; i++) {
            cout << string(colWidth, '-') << "+";
        }
        cout << endl;
    };

    printLine(n);
    cout << "| " << setw(colWidth + 1) << left << "№";
    for (int i = 0; i < n; i++) {
        cout << "| " << setw(colWidth - 1) << left << i + 1;
    }
    cout << "|" << endl;

    printLine(n);

    cout << "| " << setw(colWidth - 1) << left << "P";
    for (int num : primeNums) {
        cout << "| " << setw(colWidth - 1) << left << num;
    }
    cout << "|" << endl;
    printLine(n);

    cout << "| " << setw(colWidth + 3) << left << "Тест";
    for (bool result : primeTestAgree) {
        cout << "| " << setw(colWidth - 1) << left << (result ? "+" : "-");
    }
    cout << "|" << endl;
    printLine(n);

    cout << "| " << setw(colWidth - 1) << left << "K";
    for (int k : primeLostCount) {
        cout << "| " << setw(colWidth - 1) << left << k;
    }
    cout << "|" << endl;
    printLine(n);

}

int main() {
    srand(time(0));
    vector<int> sieve = sieveEratosthenes(500); //решето простых
    int t = 5;
    int test_k = 3;
    int k;
    int m;
    int n;
    bool isPrime;
    vector<int> primeNums; 
    vector<int> primeLostCount; 
    vector<bool> primeTestAgree; 
    for (int l = 0; l < 10; l++){
        isPrime = false;
        k = 0;
        while (!isPrime){
            m = 1;
            for (int i = 0; i < 2; i++) {
                m *= pow(sieve[rand() % sieve.size()/4], (rand() % 2) + 1); 
            }
            n = 2 * m + 1;
            if (test_Miller(n, t)) {
                primeNums.push_back(n);
                isPrime = true;
                if (probability_Test(n, test_k)){
                    primeTestAgree.push_back(true);
                }
                else{
                    primeTestAgree.push_back(false);
                }
            }
            else{
                if (probability_Test(n, test_k)){
                    k++;
                }
            }
        }
        primeLostCount.push_back(k);
    }
    cout << setw(89) << "Тест Миллера" << endl;
    printTable(primeNums, primeTestAgree, primeLostCount);
}

File: Task#2/Python/Miller.py
import random
import time

def powmod(a, x, mod):
    res = 1
    for _ in range(x):
        res = (res * a) % mod
    return res

def sieve_eratosthenes(n):
    is_prime = [True] * (n + 1)
    primes = []
    is_prime[0] = is_prime[1] = False
    
    for i in range(2, int(n**0.5) + 1):
        if is_prime[i]:
            for j in range(i * i, n + 1, i):
                is_prime[j] = False
    
    for i in range(2, n + 1):
        if is_prime[i]:
            primes.append(i)
    
    return primes

def probability_test(num, k):
    if num in (2, 3):
        return True
    if num < 2 or num % 2 == 0:
        return False
    
    s, d = 0, num - 1
    while d % 2 == 0:
        d //= 2
        s += 1
    
    for _ in range(k):
        a = random.randint(2, num - 2)
        x = powmod(a, d, num)
        y = x
        
        for _ in range(s):
            y = powmod(x, 2, num)
            if y == 1 and x not in (1, num - 1):
                return False
            x = y
        
        if y != 1:
            return False
    
    return True

def canon(n, A):
    primes = sieve_eratosthenes(500)
    q = {}
    N = n - 1
    
    for p in primes:
        R, k = N, 0
        while R % p == 0:
            R //= p
            k += 1
        if k:
            q[p] = k
    
    check = []
    for p in q:
        for a in A:
            res = powmod(a, (n - 1) // p, n)
            check.append(res)
            if res != 1:
                return 1
    
    return 0 if all(c == 1 for c in check) else 1

def test_miller(n, t):
    A = [random.randint(2, n - 2) for _ in range(t)]
    
    if sum(powmod(a, n - 1, n) for a in A) != len(A):
        return False
    
    return canon(n, A) == 1

def print_table(prime_nums, prime_test_agree, prime_lost_count):
    col_width = 12
    n = len(prime_nums)
    
    def print_line():
        print("+" + ("-" * col_width + "+") * (n + 1))
    
    print_line()
    print("|" + f" {'№':<{col_width}}" + ''.join(f"| {i+1:<{col_width-1}}" for i in range(n)) + "|")
    print_line()
    print("|" + f" {'P':<{col_width}}" + ''.join(f"| {num:<{col_width-1}}" for num in prime_nums) + "|")
    print_line()
    print("|" + f" {'Тест':<{col_width}}" + ''.join(f"| {'+' if result else '-':<{col_width-1}}" for result in prime_test_agree) + "|")
    print_line()
    print("|" + f" {'K':<{col_width}}" + ''.join(f"| {k:<{col_width-1}}" for k in prime_lost_count) + "|")
    print_line()

def main():
    random.seed(time.time())
    sieve = sieve_eratosthenes(500)
    t, test_k = 5, 3
    
    prime_nums = []
    prime_lost_count = []
    prime_test_agree = []
    
    for _ in range(10):
        is_prime = False
        k = 0
        
        while not is_prime:
            m = 1
            for _ in range(2):
                m *= pow(sieve[random.randint(0, len(sieve) // 4)], random.randint(1, 2))
            
            n = 2 * m + 1
            
            if test_miller(n, t):
                prime_nums.append(n)
                is_prime = True
                prime_test_agree.append(probability_test(n, test_k))
            else:
                if probability_test(n, test_k):
                    k += 1
        
        prime_lost_count.append(k)
    
    print("".center(89, " "), "Тест Миллера")
    print_table(prime_nums, prime_test_agree, prime_lost_count)

if __name__ == "__main__":
    main()
